server/src/managers/websocket/websocket_manager.rs

```diff
@@ -0,0 +1,366 @@
+use crate::managers::state::SignalServerState;
+use super::connection::{WSStream, WebSocketConnection, ConnectionState, ConnectionMap, ClientConnection};
+use crate::error::{ApiError, SocketManagerError};
+use crate::query::PutV1MessageParams;
+
+use chrono::Utc;
+
+/*const WS_ENDPOINTS: [&str; 24] = [
+    "v4/attachments/form/upload",
+    "v1/config",
+    "v1/certificate/delivery",
+    "v1/devices",
+    "v2/directory/auth",
+    "v1/certificate/auth/group",
+    "v1/archives/auth",
+    "v1/archives/auth/read",
+    "v1/archives/upload/form",
+    "v1/archives/media/upload/form",
+    "v1/devices/link",
+    "v1/messages",
+    "v1/messages/multi_recipient",
+    "v2/accounts/phone_number_discoverability",
+    "v1/profile",
+    "v1/archives",
+    "v1/archives/media",
+    "v1/archives/media/batch",
+    "v1/archives/media/delete",
+    "v1/devices/capabilities",
+    "v1/messages/report",
+    "v1/archives/backupid",
+    "v1/archives/keys",
+    "v1/storage/auth",
+];*/
+
+
+
+#[derive(Debug)]
+pub struct WebSocketManager<T: WSStream + Debug> {
+    sockets: ConnectionMap<T>,
+}
+
+impl<T: WSStream + Debug> Clone for WebSocketManager<T> {
+    fn clone(&self) -> Self {
+        Self {
+            sockets: Arc::clone(&self.sockets),
+        }
+    }
+}
+
+impl <T: WSStream + Debug + Send + 'static> WebSocketManager<T> {
+
+    pub fn new() -> Self {
+        Self {
+            sockets: Arc::new(Mutex::new(HashMap::new()))
+        }
+    }
+
+    pub async fn insert<U: SignalDatabase + Send + 'static>(&mut self, connection: WebSocketConnection<T>, state: SignalServerState<U>){
+        let address = connection.protocol_address();
+        let connection: ClientConnection<T> = Arc::new(Mutex::new(connection));
+        
+        self.sockets.lock().await.insert(address.clone(), connection.clone());
+        let mut mgr = self.clone();
+        
+        let protocol_address = address.clone();
+        tokio::spawn(async move {
+            let connection = connection.clone();
+            let addr = connection.lock().await.socket_address();
+            loop {
+                let mut ws_guard = connection.lock().await;
+
+                let timeout_res = tokio::time::timeout(Duration::from_millis(100), ws_guard.recv()).await;
+                let msg_opt = match timeout_res {
+                    Ok(x) => x,
+                    _ => continue // did not receive message in time, release and let others access the connection
+                };
+
+                let res = if let Some(x) = msg_opt{
+                    x
+                } else{
+                    ws_guard.close().await;
+                    break;
+                };
+                let msg = match res {
+                    Err(x) => {
+                        println!("WebSocketManager recv ERROR: {}", x);
+                        ws_guard.close().await;
+                        break;
+                    },
+                    Ok(y) => y
+                };
```

Comment:
Can't these two be combined, the println can be its own case Some(Err(x)) if necessary
```suggestion
                let msg = match msg_opt {
                    Some(Ok(y)) => y,
                    _ => {
                        ws_guard.close().await;
                        break;
                    },
                };
```

Reply:
ok